### PR TITLE
refactor: Clarify and disable unused ArgsManager flags

### PR DESCRIPTION
**Pull request by ryanofsky**
This is preparation for #16545 or another PR implementing type validation for ArgsManager settings. It fixes misleading usages of existing flags, prevents flags from being similarly misused in the future, and allows validation logic to be added without breaking backwards compatibility.

---

Currently, ALLOW_{INT|BOOL|STRING} flags don't do any real validation, so current uses of these flags are misleading and will also break backwards compatibility whenever these flags are implemented in a future PR (draft PR is #16545).

An additional complication is that while these flags don't do any real settings validation, they do affect whether setting negation syntax is allowed.

Fix this mess by disabling ALLOW_{INT|BOOL|STRING} flags until they are implemented, and adding an unambiguous DISALLOW_NEGATION flag. This is done in three commits, with the first commit cleaning up some code, the second commit adding the DISALLOW_NEGATION flag, and the next commit disabling the ALLOW_{INT|BOOL|STRING} flags.

None of the changes affect behavior in any way.

### CODE REVIEW

**Comment by ryanofsky** [comment]
Rebased 75e3ddb981f856612e1ef3ac765e2903d32781f4 -> daf070bdee9a1b4eaf3a4976537111f257bc1be6 ([`pr/argscripts.1`](https://github.com/ryanofsky/bitcoin/commits/pr/argscripts.1) -> [`pr/argscripts.2`](https://github.com/ryanofsky/bitcoin/commits/pr/argscripts.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/argscripts.1-rebase..pr/argscripts.2)) due to conflict with #22220

File: src/net_processing.cpp [hunk 1]
```diff
@@ -1300,7 +1300,7 @@ bool PeerManagerImpl::GetNodeStateStats(NodeId nodeid, CNodeStateStats& stats) c
 
 void PeerManagerImpl::AddToCompactExtraTransactions(const CTransactionRef& tx)
 {
-    size_t max_extra_txn = gArgs.GetArg("-blockreconstructionextratxn", DEFAULT_BLOCK_RECONSTRUCTION_EXTRA_TXN);
+    size_t max_extra_txn = gArgs.GetIntArg("-blockreconstructionextratxn", DEFAULT_BLOCK_RECONSTRUCTION_EXTRA_TXN);
```
Comment (kiminuo) [inline src/net_processing.cpp:1303]: I can see that once the casting is used ([here](https://github.com/bitcoin/bitcoin/pull/22766/files#diff-cafbe1353eff6084b73fd3b6c3dee603e0827348fdd2fe12dfad1e01003a84edR87)) and here it is not. Maybe it should be consistent.

File: src/bitcoin-cli.cpp [hunk 2]
```diff
@@ -807,7 +807,7 @@ static UniValue ConnectAndCallRPC(BaseRequestHandler* rh, const std::string& str
     UniValue response(UniValue::VOBJ);
     // Execute and handle connection failures with -rpcwait.
     const bool fWait = gArgs.GetBoolArg("-rpcwait", false);
-    const int timeout = gArgs.GetArg("-rpcwaittimeout", DEFAULT_WAIT_CLIENT_TIMEOUT);
+    const int timeout = gArgs.GetIntArg("-rpcwaittimeout", DEFAULT_WAIT_CLIENT_TIMEOUT);
```
Comment (kiminuo) [inline src/bitcoin-cli.cpp:810]: Should there be a cast here as `int64_t` is converted to `int`?

File: src/util/system.cpp [hunk 3]
```diff
@@ -214,30 +206,37 @@ static util::SettingsValue InterpretOption(std::string& section, std::string& ke
...
 }
 
 /**
- * Check settings value validity according to flags.
+ * Interpret settings value and apply ALLOW checking flags.
  *
- * TODO: Add more meaningful error checks here in the future
- * See "here's how the flags are meant to behave" in
- * https://github.com/bitcoin/bitcoin/pull/16097#issuecomment-514627823
+ * Return an error string and nullopt if an invalid value was provided that
```
Comment (kiminuo) [inline src/util/system.cpp:217]: Should Doxygen `@return` be used here?

File: src/util/system.h [hunk 4]
```diff
@@ -158,12 +158,18 @@ struct SectionInfo
...
+     * Flags controlling how config and command line arguments are validated and
+     * interpreted.
+     */
     enum Flags : uint32_t {
-        // Boolean options can accept negation syntax -noOPTION or -noOPTION=1
-        ALLOW_BOOL = 0x01,
-        ALLOW_INT = 0x02,
-        ALLOW_STRING = 0x04,
-        ALLOW_ANY = ALLOW_BOOL | ALLOW_INT | ALLOW_STRING,
+        ALLOW_ANY = 0x01,    //!< disable validation
```
Comment (kiminuo) [inline src/util/system.h:166]: I wonder whether `enum class Flags : uint32_t` may be used here as Jon did in https://github.com/bitcoin/bitcoin/pull/21506/ ([guideline](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Renum-class)). Would it make sense to you in this or a follow-up PR?

**Review by ryanofsky — COMMENTED** [review]
Thanks for the review!

Updated daf070bdee9a1b4eaf3a4976537111f257bc1be6 -> 7ca417e3223de2d8bf81de24e5605c8e87fcb7fe ([`pr/argscripts.2`](https://github.com/ryanofsky/bitcoin/commits/pr/argscripts.2) -> [`pr/argscripts.3`](https://github.com/ryanofsky/bitcoin/commits/pr/argscripts.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/argscripts.2..pr/argscripts.3)) updating doxygen comments as suggested

File: src/util/system.h (hunk 4 above)
Comment (kiminuo) [reply src/util/system.h:166]: I think the rationale in the guideline makes sense and that's why I suggested it. There was no specific advatange I had in mind. Anyway, it may (not) be a good follow up PR.

File: src/util/system.cpp (hunk 3 above)
Comment (kiminuo) [reply src/util/system.cpp]: It looks great.

File: src/util/system.h (hunk 4 above)
Comment (ajtowns) [reply src/util/system.h:166]: See `NetPermissionsFlags` in net_permissions.h for comparison; it gives you lots of namespacing (so you'd be writing `ArgsManager::Flags::ALLOW_BOOL` everywhere) and means you can't just treat them as ints (so bitwise or'ing things together means you have to declare your own operator)

**Review by ajtowns — COMMENTED** [review]
ACK 31a8356714821f818ef476a9581e7830dd5ed940

I found the first commit a bit hard to follow, so tried redoing your patch to understand it better. Ended up reordering it a bit, which I think turned out to make more sense (in particular it avoids adding code to set DISALLOW_NEGATION in one commit just to remove it in the next one): 

 * add DISALLOW_NEGATION to enum and check it in CheckValid
 * switch ALLOW_{INT,BOOL,STRING} to ANY or ANY|DISALLOW_NEGATION
 * split InterpretKey out of InterpretOption
 * switch InterpretOption into separate invocations of InterpretKey/InterpretValue
 * merge CheckValid into InterpretValue, adding flags,error params and returning an optional
 * rename to GetIntArgs

Branch is at https://github.com/ajtowns/bitcoin/commits/202109-disallownegation though it's not cleaned up at all. Ends up in the same state as this branch, so might be useful for other reviewers if nothing else.

(EDIT: bumped commit id)

File: src/util/system.h (hunk 4 above)
Comment (jonatack) [reply src/util/system.h:166]: Yes, two sides of a coin, I suppose.

Namespacing can allow shortening the enumerator names, since they no longer need to be unique in global namespace (and can therefore no longer need to be shouty "there can only be one"), e.g. ArgsManager::ALLOW_STRING could be ArgsManager::Flags::string. Whether that naming is better may be personal taste, but getting them out of global space seems good.

Using an enum class permits defining allowed operations: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Renum-oper and https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Renum-class ("minimize surprises: traditional enums convert to int too readily").

Comment (jonatack) [reply src/util/system.h:166]: (and changing NetPermissionFlags to an enum class found a bug for that reason.)

**Comment by MarcoFalke** [comment]
Maybe the last commit can be split up, because it is the most uncontroversial one but causes the most (silent) conflicts?

File: src/util/system.h [hunk 5]
```diff
@@ -327,7 +333,7 @@ class ArgsManager
      * @param nDefault (e.g. 1)
      * @return command-line argument (0 if invalid number) or default value
      */
-    int64_t GetArg(const std::string& strArg, int64_t nDefault) const;
+    int64_t GetIntArg(const std::string& strArg, int64_t nDefault) const;
```
[the remark this answers is not in the document]
Comment (ajtowns) [reply src/util/system.h:336]: I think I finally figured out how you can go a little bit further than the above, ending up with something like:

```c++
struct NetSettings
{
    int64_t blockreconstructionextratxn;
    int64_t maxorphantx;
    bool capturemessages;

    template<typename C, typename... Args>
    static inline void F(Args&... args) {
        return C::Do(args...,
            C::Defn( &NetSettings::blockreconstructionextratxn, "-blockreconstructionextratxn=<n>", strprintf("Extra transactions to keep in memory for compact block reconstructions (default: %u)", DEFAULT_BLOCK_RECONSTRUCTION_EXTRA_TXN), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS, DEFAULT_BLOCK_RECONSTRUCTION_EXTRA_TXN ),
            C::Defn( &NetSettings::maxorphantx, "-maxorphantx=<n>", strprintf("Keep at most <n> unconnectable transactions in memory (default: %u)", DEFAULT_MAX_ORPHAN_TRANSACTIONS), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS, DEFAULT_MAX_ORPHAN_TRANSACTIONS ),
            C::Defn( &NetSettings::capturemessages, "-capturemessages", "Capture all P2P messages to disk", ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST, false )
        );
    }
};

void RegisterNetProcessingArgs(ArgsManager& args)
{
    SettingsRegister<NetSettings>::Register(args);
}

NetSettings GetNetSettings(const ArgsManager& args)
{
    return SettingsRegister<NetSettings>::Get(args);
}

class PeerManagerImpl
{
private:
    const NetSettings m_settings;
    PeerManagerImpl(..., const ArgsManager& args) : m_settings{GetNetSettings(args)), ... { ... }
    ...
};
```

The idea being that this way:
 * it can infer the argument type directly from the type of the struct member so that you can't accidentally specify different types between `args.AddArg<int>` and `Get<bool>`
  * that the settings are const at runtime so can be accessed without any additional locks
  * you only have to access the ArgsManager (and do string parsing) at setup time
  * you don't have to make up lots of new names for everything or add too much boilerplate

Branch built on top of this PR that has the above working at https://github.com/ajtowns/bitcoin/tree/202109-settings_struct

**Comment by hebasto** [comment]
Concept ACK.

Wondering what are benefits of the `DISALLOW_NEGATION` flag in comparison to the `ALLOW_NEGATED` one (not an issue at all, just trying to understand motivation)?

File: src/util/system.h [hunk 6]
```diff
@@ -158,12 +158,18 @@ struct SectionInfo
...
-        ALLOW_STRING = 0x04,
-        ALLOW_ANY = ALLOW_BOOL | ALLOW_INT | ALLOW_STRING,
+        ALLOW_ANY = 0x01,    //!< disable validation
+        ALLOW_BOOL = 0x02,   //!< unimplemented, draft implementation in #16545
+        ALLOW_INT = 0x04,    //!< unimplemented, draft implementation in #16545
+        ALLOW_STRING = 0x08, //!< unimplemented, draft implementation in #16545
+        ALLOW_LIST = 0x10,   //!< unimplemented, draft implementation in #16545
+        DISALLOW_NEGATION = 0x20, //! disallow -nofoo syntax
+
         DEBUG_ONLY = 0x100,
```
Comment (hebasto) [inline src/util/system.h]: 1) s/`ALLOW_ANY`/`DISABLE_VALIDATION`/, and drop the `//!< disable validation` comment?
2) as `ALLOW_ANY` is not a combination of other flags now, would it simplify the code if `ALLOW_ANY = 0`? (going to check this suggestion later)

nits:
1) could we switch values to "shifted ones" `(1 << N)` for better readability?
2) assuming that another type could be added in the future, e.g., `ALLOW_PAIR`, maybe `DISALLOW_NEGATION = 0x80` or `DISALLOW_NEGATION = (1 << 30)`?
3) `clang-format` complains about not justified comments

File: src/util/system.cpp [hunk 7]
```diff
@@ -214,30 +210,43 @@ static util::SettingsValue InterpretOption(std::string& section, std::string& ke
...
  */
-static bool CheckValid(const std::string& key, const util::SettingsValue& val, unsigned int flags, std::string& error)
-{
-    if (val.isBool() && !(flags & ArgsManager::ALLOW_BOOL)) {
-        error = strprintf("Negating of -%s is meaningless and therefore forbidden", key);
+static std::optional<util::SettingsValue> InterpretValue(const std::string& key,
+    const std::string& value,
+    bool negated,
+    unsigned int flags,
+    std::string& error)
```
Comment (hebasto) [inline src/util/system.cpp:234]: nit: to make `clang-format` happy:
```suggestion
static std::optional<util::SettingsValue> InterpretValue(
    const std::string& key,
    const std::string& value,
    bool negated,
    unsigned int flags,
    std::string& error)
```

**Review by hebasto — COMMENTED** [review]
Approach ACK 31a8356714821f818ef476a9581e7830dd5ed940.

I really like @ajtowns's suggestions:
- to [split the first commit](https://github.com/bitcoin/bitcoin/pull/22766#pullrequestreview-743341262)
- to [replace a returned pure `bool` with a structure](https://github.com/bitcoin/bitcoin/pull/22766#discussion_r699851892)

**Review by ryanofsky — COMMENTED** [review]
Updated 31a8356714821f818ef476a9581e7830dd5ed940 -> 0768cdc3e77f9670f0115cdf197a282c355bc287 ([`pr/argscripts.4`](https://github.com/ryanofsky/bitcoin/commits/pr/argscripts.4) -> [`pr/argscripts.5`](https://github.com/ryanofsky/bitcoin/commits/pr/argscripts.5), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/argscripts.4..pr/argscripts.5)) making suggested changes. Biggest change is dropping the last commit.

---

re: https://github.com/bitcoin/bitcoin/pull/22766#issuecomment-910001542

> Maybe the last commit can be split up, because it is the most uncontroversial one but causes the most (silent) conflicts?

Thanks, dropped last commit from this PR and posted as #22976.

---

re: https://github.com/bitcoin/bitcoin/pull/22766#pullrequestreview-743341262

> Branch is at [ajtowns/bitcoin@`202109-disallownegation` (commits)](https://github.com/ajtowns/bitcoin/commits/202109-disallownegation) though it's not cleaned up at all. Ends up in the same state as this branch, so might be useful for other reviewers if nothing else.

Thanks! I couldn't use the branch exactly because scripted-diff b82541a5f48ffb47d8b1bdc161c14956b0435502 was not valid, but I did split up the first commit basically using your approach and tagged you as co-author.


---

re: https://github.com/bitcoin/bitcoin/pull/22766#issuecomment-912284343

> Wondering what are benefits of the `DISALLOW_NEGATION` flag in comparison to the `ALLOW_NEGATED` one (not an issue at all, just trying to understand motivation)?

There are two types of options:

- Typed options (int, string, bool, list options with implementation under discussion in #16545)
- Untyped options (ALLOW_ANY options)

For untyped options, I think DISALLOW_NEGATION is better than ALLOW_NEGATED because negation has always been allowed for these options. For the 200 or so existing untyped options that allow negation, this PR is leaving them untouched, not changing their behavior or the way they are declared. There are 5 untyped options that don't allow negation. This PR is changing the declaration of these 5 options to be less misleading and actually match their implemented behavior.

For typed options, I think DISALLOW_NEGATION should rarely or never be used. Almost all existing options accept negation. Negation obviously makes sense for bool options. Existing usage shows it also makes sense for most int and string options. For some int and string options, it doesn't make sense, but in these cases generic negation handling and the generic negation message "Negating of -xxx is meaningless and therefore forbidden" is worse than just writing validation logic that has to be written anyway and giving specific errors like "Setting -xxx value false is out of range, the acceptable range is YYY-ZZZ." "Setting -xxx value false is not recognized, acceptable values are A, B, and C."

If I'm wrong about this and the DISALLOW_NEGATION flag ever starts providing more useful messages, or starts being used many more places, it can be enabled by default in the future. But don't think these things will happen and I think even if they did happen later it would be premature to disallow negation by default and update 200 existing option declarations right now.

---

re: https://github.com/bitcoin/bitcoin/pull/22766#pullrequestreview-745835425

> I really like @ajtowns's suggestions:
>
> * to [split the first commit](https://github.com/bitcoin/bitcoin/pull/22766#pullrequestreview-743341262)
> * to [replace a returned pure `bool` with a structure](https://github.com/bitcoin/bitcoin/pull/22766#discussion_r699851892)

Thanks, implemented both.

**Comment by ryanofsky** [comment]
Rebased 0768cdc3e77f9670f0115cdf197a282c355bc287 -> 5c04ca81fd2c44fed6487123500b49434280d514 ([`pr/argscripts.5`](https://github.com/ryanofsky/bitcoin/commits/pr/argscripts.5) -> [`pr/argscripts.6`](https://github.com/ryanofsky/bitcoin/commits/pr/argscripts.6), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/argscripts.5-rebase..pr/argscripts.6)) due to silent conflict with #23025 in scripted diff commit

**Review by promag — COMMENTED** [review]
Concept ACK, code review on 1st commit. I need to understand the big picture from #16545 in order to review the remaining changes here.

**Comment by ryanofsky** [comment]
Thanks for review! Just wanted to quickly say you do not need to care anything about #16545 for this PR to make sense. This PR is disabling flags `ALLOW_INT/STRING/BOOL` which are broken, by commenting them out. #16545 is reenabling these flags by uncommenting them again and implementing them.

Even if #16545 is rejected, these flags are misleading now and should be disabled. The only effect these flags have now is to disallow negation, so this PR is replacing them with a clearer `DISALLOW_NEGATION` flag. I expect this flag to be rarely used in the future, and probably to be removed later. Definite it here is just useful to make current code clear and avoid changing behavior in any way with this PR.

**Review by promag — COMMENTED** [review]
Code review ACK 5c04ca81fd2c44fed6487123500b49434280d514.

Thanks for the clarification. Indeed, the existing flags are misleading when in practice they are used to infer if negation is accepted. I got distracted by `ALLOW_LIST`, just realized it's only added in #16545.

**Review by kiminuo — COMMENTED** [review]
Concept ACK

**Review by ryanofsky — COMMENTED** [review]
Thanks for review! Updated just making suggested changes.

Updated 5c04ca81fd2c44fed6487123500b49434280d514 -> b8fd5d54fafe94cb40a43e3e1366a6855a512e4e ([`pr/argscripts.6`](https://github.com/ryanofsky/bitcoin/commits/pr/argscripts.6) -> [`pr/argscripts.7`](https://github.com/ryanofsky/bitcoin/commits/pr/argscripts.7), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/argscripts.6..pr/argscripts.7)) with suggested changes

**Review by ryanofsky — COMMENTED** [review]
re: https://github.com/bitcoin/bitcoin/pull/22766#discussion_r730275392

> nit: As there are many "keys" in Bitcoin codebase and the word meaning is heavily overloaded, maybe it would make sense to have `SettingKeyInfo` / `SettingKey` / `ArgKeyInfo` or something.

EDIT: moved reply to https://github.com/bitcoin/bitcoin/pull/22766#discussion_r730282231

**Review by promag — COMMENTED** [review]
Code review ACK b8fd5d54fafe94cb40a43e3e1366a6855a512e4e.

**Review by promag — COMMENTED** [review]
Code review ACK c5d7e34bd9a4ad752c5ec88032420e2e90ab17ab, which as the new argument `-legacy`.
